dipy/viz/ui.py

```diff
@@ -0,0 +1,202 @@
+import numpy as np
+
+from dipy.viz.interactor import CustomInteractorStyle
+
+from dipy.utils.optpkg import optional_package
+
+# Allow import, but disable doctests if we don't have vtk.
+vtk, have_vtk, setup_module = optional_package('vtk')
+
+if have_vtk:
+    version = vtk.vtkVersion.GetVTKSourceVersion().split(' ')[-1]
+    major_version = vtk.vtkVersion.GetVTKMajorVersion()
+else:
+    vtkInteractorStyleUser = object
+
+
+class UI(object):
+    """ An umbrella class for all UI elements.
+    While adding UI elements to the renderer, we need to go over all the
+     sub-elements that come with it and add those to the renderer too.
+    There are several features that are common to all the UI elements:
+    - ui_param : This is an attribute that can be passed to the UI object
+                by the interactor. Thanks to Python's dynamic type-setting
+                this parameter can be anything.
+    - ui_list : This is used when there are more than one UI elements inside
+               a UI element. Inside the renderer, they're all iterated and added.
+    - parent_UI: This is useful of there is a parent UI element and its reference
+                needs to be passed down to the child.
+    """
+    def __init__(self):
+        self.ui_param = None
+        self.ui_list = list()
+
+        self.parent_UI = None
+        self._callbacks = []
+
+    def get_actors(self):
+        """ Returns the actors that compose this UI component. """
+        msg = "Subclasses of UI must implement `get_actors(self)`."
+        raise NotImplementedError(msg)
+
+    def add_to_renderer(self, ren):
+        """ Allows UI objects to add their own props to the renderer.
+        Parameters
+        ----------
+        ren : renderer
+        """
+        ren.add(*self.get_actors())
+
+        # Get a hold on the current interactor style.
+        iren = ren.GetRenderWindow().GetInteractor().GetInteractorStyle()
+
+        for callback in self._callbacks:
+            if not isinstance(iren, CustomInteractorStyle):
```

Comment:
Yep, this has been tested. 
https://github.com/ranveeraggarwal/dipy/blob/ui/dipy/viz/tests/test_ui.py#L155